### ISSUE TITLE
Panel apps not working with Bokeh 2.0

### ISSUE THREAD

**Issue by sdc50** [tethysplatform/tethys#540]
The bokeh_tutorial app has both a Bokeh app and a Panel app example. I can get both to work with:
- Bokeh 1.4
- Panel 0.6.4
But any other version of Panel causes the Panel app to break. Even with the recent PR that updates compatibility with Bokeh 2.0, I'm still getting the error:

Environment (```conda list```):

<details>

```
(tethys-bokeh2)$ conda list
# packages in environment at tethys-bokeh2:
#
# Name                    Version                   Build  Channel
appdirs                   1.4.3                      py_1    conda-forge
asgiref                   3.2.5                      py_0    conda-forge
asn1crypto                1.3.0                    py37_0    conda-forge
asynctest                 0.12.2                     py_0    conda-forge
attrs                     19.3.0                     py_0    conda-forge
autobahn                  20.2.2                     py_0    conda-forge
automat                   20.2.0                     py_0    conda-forge
bcrypt                    3.1.7            py37h0b31af3_0    conda-forge
beautifulsoup4            4.8.2                    py37_0    conda-forge
blinker                   1.4                        py_1    conda-forge
bokeh                     2.0.0            py37hc8dfbb8_0    conda-forge
bumpversion               0.5.3                   py_1001    conda-forge
bzip2                     1.0.8                h0b31af3_2    conda-forge
ca-certificates           2019.11.28           hecc5488_0    conda-forge
certifi                   2019.11.28       py37hc8dfbb8_1    conda-forge
cffi                      1.14.0           py37h356ff06_0    conda-forge
cftime                    1.1.1            py37h10e2902_0    conda-forge
channels                  2.4.0                      py_0    tethysplatform/label/dev
chardet                   3.0.4           py37hc8dfbb8_1006    conda-forge
click                     7.1.1              pyh8c360ce_0    conda-forge
cloudpickle               1.3.0                      py_0    conda-forge
conda                     4.8.3            py37hc8dfbb8_1    conda-forge
conda-package-handling    1.6.0            py37h9bfed18_2    conda-forge
condorpy                  0.5.3                      py_0    tethysplatform/label/dev
constantly                15.1.0                     py_0    conda-forge
coverage                  5.0.4            py37h9bfed18_0    conda-forge
cryptography              2.8              py37he655712_2    conda-forge
curl                      7.68.0               h8754def_0    conda-forge
cytoolz                   0.10.1           py37h0b31af3_0    conda-forge
daphne                    2.4.1                      py_0    tethysplatform/label/dev
dask                      1.2.2                      py_3    conda-forge
dask-core                 1.2.2                      py_0    conda-forge
defusedxml                0.6.0                      py_0    conda-forge
distributed               1.28.1                   py37_0    conda-forge
distro                    1.4.0                      py_0    conda-forge
django                    2.2.8                    py37_0    conda-forge
django-analytical         2.5.0                      py_0    tethysplatform/label/dev
django-bootstrap3         11.1.0                     py_0    conda-forge
django-environ            0.4.5                      py_1    conda-forge
django-filter             2.2.0                      py_0    conda-forge
django-gravatar2          1.4.3                      py_0    tethysplatform/label/dev
django-guardian           2.1.0                      py_0    conda-forge
django-model-utils        4.0.0                      py_0    conda-forge
django-ranged-response    0.2.0                      py_0    tethysplatform/label/dev
django-recaptcha2         1.4.1                      py_0    tethysplatform/label/dev
django-session-security   2.6.5                      py_0    tethysplatform/label/dev
django-simple-captcha     0.5.12                     py_0    tethysplatform/label/dev
django-termsandconditions 2.0                        py_0    tethysplatform/label/dev
djangorestframework       3.11.0                     py_0    conda-forge
docker-py                 4.2.0                    py37_0    conda-forge
docker-pycreds            0.4.0                      py_0    conda-forge
entrypoints               0.3             py37hc8dfbb8_1001    conda-forge
expat                     2.2.9                h4a8c4bd_2    conda-forge
factory_boy               2.11.1                     py_0    conda-forge
faker                     4.0.2                      py_0    conda-forge
flake8                    3.7.9            py37hc8dfbb8_1    conda-forge
freetype                  2.10.1               h8da9a1a_0    conda-forge
future                    0.18.2           py37hc8dfbb8_1    conda-forge
geoalchemy2               0.6.3                      py_0    conda-forge
gettext                   0.19.8.1          h46ab8bc_1002    conda-forge
git                       2.25.0          pl526hdc91d69_0    conda-forge
gsconfig                  1.1.0                      py_0    tethysplatform/label/dev
hdf4                      4.2.13            h84186c3_1003    conda-forge
hdf5                      1.10.5          nompi_h3e39495_1104    conda-forge
heapdict                  1.0.1                      py_0    conda-forge
hs_restclient             1.3.5                      py_0    conda-forge
hyperlink                 17.3.1                     py_0    conda-forge
icu                       64.2                 h6de7cb9_1    conda-forge
idna                      2.9                        py_1    conda-forge
incremental               17.5.0                     py_0    conda-forge
jinja2                    2.11.1                     py_0    conda-forge
jpeg                      9c                h1de35cc_1001    conda-forge
krb5                      1.16.4               h1752a42_0    conda-forge
libblas                   3.8.0               16_openblas    conda-forge
libcblas                  3.8.0               16_openblas    conda-forge
libcurl                   7.68.0               h709d2b2_0    conda-forge
libcxx                    9.0.1                         1    conda-forge
libedit                   3.1.20170329      hcfe32e1_1001    conda-forge
libffi                    3.2.1             h4a8c4bd_1007    conda-forge
libgfortran               4.0.0                         2    conda-forge
libiconv                  1.15              h0b31af3_1006    conda-forge
liblapack                 3.8.0               16_openblas    conda-forge
libnetcdf                 4.7.3           nompi_hda4e5f1_101    conda-forge
libopenblas               0.3.9                h3d69b6c_0    conda-forge
libpng                    1.6.37               hbbe82c9_1    conda-forge
libpq                     12.2                 h554dc5a_0    conda-forge
libprotobuf               3.11.4               hd174df1_0    conda-forge
libssh2                   1.8.2                hcdc9a53_2    conda-forge
libtiff                   4.1.0                h2ae36a8_6    conda-forge
libwebp-base              1.1.0                         2    conda-forge
libxml2                   2.9.10               h53d96d6_0    conda-forge
llvm-openmp               9.0.1                h28b9765_2    conda-forge
locket                    0.2.0                      py_2    conda-forge
lz4-c                     1.8.3             h6de7cb9_1001    conda-forge
markdown                  3.2.1                      py_0    conda-forge
markupsafe                1.1.1            py37h9bfed18_1    conda-forge
mccabe                    0.6.1                      py_1    conda-forge
mock                      3.0.5            py37hc8dfbb8_1    conda-forge
msgpack-python            0.6.2            py37ha1b3eb9_0    conda-forge
ncurses                   6.1               h0a44026_1002    conda-forge
netcdf4                   1.5.3           nompi_py37hcf2264a_102    conda-forge
numpy                     1.18.1           py37hde6bac1_0    conda-forge
oauthlib                  3.0.1                      py_0    conda-forge
olefile                   0.46                       py_0    conda-forge
openssl                   1.1.1e               h0b31af3_0    conda-forge
owslib                    0.19.2                     py_0    conda-forge
packaging                 20.1                       py_0    conda-forge
pandas                    1.0.3            py37h94625e5_0    conda-forge
panel                     0.9.1                      py_0    conda-forge
param                     1.9.3                      py_0    conda-forge
paramiko                  2.7.1                    py37_0    conda-forge
partd                     1.1.0                      py_0    conda-forge
pbr                       5.4.2                      py_0    conda-forge
pcre                      8.44                 h4a8c4bd_0    conda-forge
perl                      5.26.2            haec8ef5_1006    conda-forge
pillow                    7.0.0            py37hfd78ece_1    conda-forge
pip                       20.0.2                     py_2    conda-forge
plotly                    4.5.4              pyh8c360ce_0    conda-forge
postgresql                12.2                 h16d8c28_0    conda-forge
proj                      6.3.1                h773a61f_1    conda-forge
protobuf                  3.11.4           py37h4a8c4bd_0    conda-forge
psutil                    5.7.0            py37h9bfed18_1    conda-forge
psycopg2                  2.8.4            py37hafa8578_1    conda-forge
pyasn1                    0.4.8                      py_0    conda-forge
pyasn1-modules            0.2.7                      py_0    conda-forge
pycodestyle               2.5.0                      py_0    conda-forge
pycosat                   0.6.3           py37h9bfed18_1004    conda-forge
pycparser                 2.20                       py_0    conda-forge
pycrypto                  2.6.1           py37h9bfed18_1004    conda-forge
pyct                      0.4.6                      py_0    conda-forge
pyct-core                 0.4.6                      py_0    conda-forge
pyflakes                  2.1.1                      py_0    conda-forge
pyhamcrest                2.0.2                      py_0    conda-forge
pyjwt                     1.7.1                      py_0    conda-forge
pynacl                    1.3.0           py37h0b31af3_1001    conda-forge
pyopenssl                 19.1.0                     py_1    conda-forge
pyparsing                 2.4.6                      py_0    conda-forge
pyproj                    2.6.0            py37hf6d6e9e_0    conda-forge
pysocks                   1.7.1            py37hc8dfbb8_1    conda-forge
python                    3.7.6           hfa4aa89_4_cpython    conda-forge
python-dateutil           2.8.1                      py_0    conda-forge
python3-openid            3.1.0           py37hc8dfbb8_1002    conda-forge
python_abi                3.7                     1_cp37m    conda-forge
pytz                      2019.3                     py_0    conda-forge
pyviz_comms               0.7.4              pyh8c360ce_0    conda-forge
pyyaml                    5.3.1            py37h9bfed18_0    conda-forge
readline                  8.0                  hcfe32e1_0    conda-forge
requests                  2.23.0             pyh8c360ce_2    conda-forge
requests-mock             1.7.0                      py_0    conda-forge
requests-oauthlib         1.2.0                      py_0    conda-forge
requests-toolbelt         0.9.1                      py_0    conda-forge
retrying                  1.3.3                      py_2    conda-forge
ruamel_yaml               0.15.80         py37h9bfed18_1001    conda-forge
scp                       0.13.2                     py_0    conda-forge
selenium                  3.141.0         py37h9bfed18_1001    conda-forge
service_identity          18.1.0                     py_0    conda-forge
setuptools                46.0.0           py37hc8dfbb8_2    conda-forge
siphon                    0.8.0                 py37_1002    conda-forge
six                       1.14.0                     py_1    conda-forge
social-auth-app-django    3.1.0                      py_0    conda-forge
social-auth-core          3.3.0            py37hc8dfbb8_2    conda-forge
sortedcontainers          2.1.0                      py_0    conda-forge
soupsieve                 1.9.4            py37hc8dfbb8_1    conda-forge
sqlalchemy                1.3.15           py37h9bfed18_1    conda-forge
sqlite                    3.30.1               h93121df_0    conda-forge
sqlparse                  0.3.1                      py_0    conda-forge
tblib                     1.6.0                      py_0    conda-forge
tethys-platform           3.1.0             dev3_g947dc53    tethysplatform/label/dev
tethys_dask_scheduler     1.0.2                      py_0    tethysplatform/label/dev
tethys_dataset_services   2.0.1                      py_0    tethysplatform/label/dev
tethysapp-bokeh-tutorial  0.0.1                     dev_0    <develop>
text-unidecode            1.3                        py_0    conda-forge
tk                        8.6.10               hbbe82c9_0    conda-forge
toolz                     0.10.0                     py_0    conda-forge
tornado                   6.0.4            py37h9bfed18_1    conda-forge
tqdm                      4.43.0                     py_0    conda-forge
twisted                   20.3.0           py37h9bfed18_0    conda-forge
txaio                     20.1.1                     py_0    conda-forge
typing_extensions         3.7.4.1          py37hc8dfbb8_1    conda-forge
tzcode                    2019a             h01d97ff_1002    conda-forge
unidecode                 1.1.1                      py_0    conda-forge
urllib3                   1.25.7           py37hc8dfbb8_1    conda-forge
websocket-client          0.57.0           py37hc8dfbb8_1    conda-forge
wheel                     0.34.2                     py_1    conda-forge
xz                        5.2.4             h0b31af3_1002    conda-forge
yaml                      0.2.2                h0b31af3_1    conda-forge
zict                      2.0.0                      py_0    conda-forge
zlib                      1.2.11            h0b31af3_1006    conda-forge
zope.interface            5.0.0            py37h9bfed18_0    conda-forge
zstd                      1.4.4                hed8d7c8_2    conda-forge

```
</details>

Error:

```
Error rendering Bokeh items: Error: Model 'panel.models.markup.HTML' does not exist. This could be due to a widget or a custom model not being registered before first usage.
```

I've tried this on both Mac OS and Linux, but still get the same behavior.

**Comment by sdc50** [comment]
@msouff You said that you were able to get this to work with Bokeh 2.0 + Panel 0.9.1. I'm having no luck. Is there something I'm missing? Did you have to make any changes to the bokeh_tutorial app when updating to Bokeh 2.0?

**Comment by swainn** [comment]
@sdc50 are you having this issue in the current release version? Or is this only an issue in dev?

**Comment by msouff** [comment]
> @msouff You said that you were able to get this to work with Bokeh 2.0 + Panel 0.9.1. I'm having no luck. Is there something I'm missing? Did you have to make any changes to the bokeh_tutorial app when updating to Bokeh 2.0?

Just for the record, this is not working for me either.

**Comment by sdc50** [comment]
@philippjfr walked me through getting this working in Tethys with the Bokeh=2.0.0 and Panel=0.9.5 and the following modifications to Tethys:

In `tethys_apps/base/app_base.py`:
```
- from tethys_portal.consumers.bokeh_consumers import BokehAutoloadJsCDN
+ from bokeh.server.django.consumers import AutoloadJsConsumer

...

-        http_url = url(urlpattern('/autoload.js'), BokehAutoloadJsCDN,
+        http_url = url(urlpattern('/autoload.js'), AutoloadJsConsumer,
```

in `tethys_portal/settings.py`

```
+ from bokeh.settings import settings
+ settings.resources = 'cdn'
```

I'm not sure whether or not this will work with Bokeh=2.0.1 (See https://github.com/holoviz/panel/issues/1205), but will test soon.

@msouff it seems like with Bokeh 2.0 we eliminate the need for `BokehAutoloadJsCDN`, but I'm not sure of all of the reasons why we needed it in the first place. What do you think?

**Comment by msouff** [comment]
@sdc50, the main reason was the cdn. I did try using the autoload consumer directly earlier but I did not know you could do this:

> + from bokeh.settings import settings
> + settings.resources = 'cdn'

I can't remember another reason for having our own bokeh consumer other than the cdn, and just using theirs would be way easier for testing and future updates.

**Comment by philippjfr** [comment]
I think this only became possible quite recently so it makes sense that you didn't do this initially. Switching now would definitely be preferable because it will certainly simplify maintenance going forward.

**Comment by swainn** [comment]
@sdc50 @msouff I merged #546 related to this into master. Do we need to also do a hotfix?

**Comment by msouff** [comment]
@swainn  I thought we did. Otherwise, when someone tries to use Panel it will not work. Since Panel is not part of tethys they will probably be using the latest version, which doesn't work with bokeh 1.4.

**Comment by swainn** [comment]
@msouff The PR you submitted was to the master branch. So those changes will not be in effect until 3.1.0 is released. If you want to do a hotfix, then you need to pull release, cherry-pick the changes to that branch (DO NOT MERGE FROM MASTER), and submit a PR to the release branch. However, I'm not sure this is an issue in the release version, b/c we locked down the versions of dependencies in our environment.yml for release.

**Comment by swainn** [comment]
@msouff I enabled tests on the release branch and discovered that the Bokeh tests were failing there too. Submitted PR #556 to pin it to version 1.4.0 as an attempt to fix it. We'll get your fix when 3.1 is released.
